### DISCUSSION TITLE
How can you get the sourcemaps data in a parcel-optimizer?

### DISCUSSION THREAD

**Discussion by wis** [parcel-bundler/parcel#5738] [Q&A] [answered]
I'm trying to write a parcel-optimizer that appends the sourcemaps data to the end of JS files (in an inline sourcemaps, via data:-URIs), something like
`//@ sourceMappingURL=data:application/json;base64,eyJ2ZXJza...`
or
`//@ sourceMappingURL=data:application/json;charset=utf-8;base64,eyJ2ZXJza...`

so my question is, in which stage of the pipeline are the sourcemaps generated? and can you get hold of them in a parcel-optimizer?

The reason I need to build an parcel-optimizer that does this is because of this Chromium  [issue](https://bugs.chromium.org/p/chromium/issues/detail?id=974543) and the suggestion as a solution in [this comment specifically](https://bugs.chromium.org/p/chromium/issues/detail?id=974543#c28).

text of the comment: 
>Re comment 25: Can you create a separate bug for caching problem?
>
> With the DWARF C++ Extension, we've been successfully using **inline sourcemaps (via data:-URIs)**, which get's around both the chrome-extension:-URI limitation and also avoids the caching problem (which independently needs to be fixed). **Have you considered just using data:-URIs** for SWEB-react?

by the way, I already built and do use [a custom parcel-optimizer](https://github.com/wis/SWEB-react/tree/master/packages/parcel-optimizer-relative-to-local-server-sourcemappingurl-rewriter) that rewrites the last line to link to a local server hosted URL instead of a relative one for the sourcemaps.

**Comment by devongovett** [comment, marked as answer]
Parcel has a built in way to inline source maps. You can configure your target to do so in package.json:

```json
"targets": {
    "default": {
      "sourceMap": {
        "inline": true,
        "inlineSources": true
      }
    }
}
```

Does this work for your use case?

**Reply by wis** [reply]
can you configure it to emit Data URIs like I showed above?
I need to specifically use Data URIs to circumvent devtools bugs with Extension development in Chromium and Firefox related to fetching non-http schemed URLs of sourcemaps and their caching.

**Reply by mischnic** [reply]
`"inline": true` = data-urls

https://v2.parceljs.org/configuration/package-json/#sourcemap

**Reply by wis** [reply]
oh, great, thanks guys I'll check it out.
does HMR still work when inlining sourcemaps via Data URIs?

**Reply by mischnic** [reply]
They are broken with HMR anyway: https://github.com/parcel-bundler/parcel/issues/643
